### PR TITLE
Harmony 1090 close scroll sessions

### PR DESCRIPTION
**Pull request by indiejames**
Closes CMR scroll session when a job is canceled or the query-cmr tasks complete or fail

### CODE REVIEW

**Review by chris-durbin — APPROVED** [review]
Tested this out in the sandbox with a workload run and verified we did not run into any CMR 429 errors due to too many scroll contexts opened. Verified via the logs that we were calling clear scrolling session for each scroll initiated.

Just some minor comments and trying to figure out why there are so many fixtures added for a null scroll-id.

File: app/backends/workflow-orchestration.ts [hunk 1]
```diff
@@ -286,11 +294,10 @@ export async function updateWorkItem(req: HarmonyRequest, res: Response): Promis
         workItem.workflowStepIndex,
         SUCCESSFUL_WORK_ITEM_STATUSES,
       );
-      const thisStep = await getWorkflowStepByJobIdStepIndex(
-        tx,
-        workItem.jobID,
-        workItem.workflowStepIndex,
-      );
+
+      if (isQueryCmr && successWorkItemCount === thisStep.workItemCount) {
```
[the remark this answers is not in the document]
Comment (indiejames) [reply app/backends/workflow-orchestration.ts:298]: right

File: fixtures/cmr.uat.earthdata.nasa.gov-443/164693196161939791 [hunk 2]
```diff
@@ -0,0 +1,25 @@
+POST /search/clear-scroll
```
[the remark this answers is not in the document]
Comment (indiejames) [reply fixtures/cmr.uat.earthdata.nasa.gov-443/164693196161939791:1]: removed

File: app/util/cmr.ts [hunk 3]
```diff
@@ -515,10 +548,32 @@ export function initiateGranuleScroll(
...
+ * 
+ * @param scrollID - the scroll-id of the scroll session
+ */
+export async function clearScrollSession(scrollId: string): Promise<void> {
+  logger.debug(`Clearing scroll session for scroll-id: ${scrollId}`);
+  if (scrollId) {
+    try {
+      await _cmrPostBody('/search/clear-scroll', { scroll_id: scrollId });
+    } catch {
+      // Do nothing - CMR will close the scroll session after ten minutes anyway. `cmrPostBody`
```
[the remark this answers is not in the document]
Comment (indiejames) [reply app/util/cmr.ts:573]: Done

File: app/models/work-item.ts [hunk 4]
```diff
@@ -498,3 +498,21 @@ export async function workItemCountByServiceIDAndStatus(
...
+ * 
+ * @param tx - the transaction to use for querying
+ * @param jobID - the JobID
+ * @returns A promise containing a scroll-id or null if the job does not use query-cmr
+ */
+export async function getScrollIdForJob(
+  tx: Transaction,
+  jobID: string,
+): Promise<string> {
+  const workItems = await getWorkItemsByJobIdAndStepIndex(tx, jobID, 1);
```
[the remark this answers is not in the document]
Comment (indiejames) [reply app/models/work-item.ts:513]: Done.

File: fixtures/cmr.uat.earthdata.nasa.gov-443/16469319511057616 [hunk 5]
```diff
@@ -0,0 +1,25 @@
+POST /search/clear-scroll
```
[the remark this answers is not in the document]
Comment (indiejames) [reply fixtures/cmr.uat.earthdata.nasa.gov-443/16469319511057616]: Done.

**Review by vinnyinverso — APPROVED** [review]

**Pull request merged**